### PR TITLE
fix(dbless): make uniqueness checks workspace-aware

### PR DESCRIPTION
**Pull request by hishamhm**
No tests here because I don't think we currently have workspace tests in the OSS dbless test suite.

### CODE REVIEW

**Review by outsinre — APPROVED** [review]

**Review by windmgc — APPROVED** [review]

File: kong/db/schema/others/declarative_config.lua [hunk 1]
```diff
@@ -389,9 +414,9 @@ local function populate_references(input, known_entities, by_id, by_key, expecte
 end
 
 
-local function find_entity(key, entity, by_key, by_id)
-  return (by_key[entity] and by_key[entity][key])
-      or (by_id[entity]  and by_id[entity][key])
+local function find_entity(ws_id, key, entity, by_key, by_id)
+  return (by_key[ws_id] and by_key[ws_id][entity] and by_key[ws_id][entity][key])
+      or (by_id[entity] and by_id[entity][key])
```
Comment (gszr) [inline kong/db/schema/others/declarative_config.lua:419]: Maybe a corner case we don't care about (or a misunderstanding of mine), but would the same issue happen with entity ID if the user chose to assign the same ID for two entities in separate workspaces?

**Review by gszr — COMMENTED** [review]
> No tests here because I don't think we currently have workspace tests in the OSS dbless test suite.

@hishamhm are there related tests on the EE codebase?

File: kong/db/schema/others/declarative_config.lua (hunk 1 above)
Comment (hishamhm) [reply kong/db/schema/others/declarative_config.lua:419]: It's a good question, I had to double-check, and the answer is "yes but no". At the moment, what happens is that ids are being checked for uniqueness across workspaces, so this is now _more_ strict than DB-mode (which looks like it has `("ws_id", "id")` UNIQUE constraints; tbh I don't recall why). In practice, people do expect UUIDs to be, well, Universally Unique, so I don't expect people considering the lack of support for repeated ids across workspaces to be a bug. In fact, in an ideal world the uniqueness constraint should have been across all entities, but our data model does not enforce that all entities must have an `id` field (even though parts or our DAO system does!)

In short, the problem that this PR fixed was that the checks were too strict for endpoint keys, but I don't think they're too strict for ids.

A reassuring thing is that we're not losing input with the extra strictness (which was the case before the initial dbless uniqueness PR). If someone does use non-unique ids, they will get a graceful error message.

**Comment by hishamhm** [comment]
> @hishamhm are there related tests on the EE codebase?

Related yes, but not specific tests, I think. There are tests that rely on this fix, so they effectively function as regression tests, but we could use some explicit cross-workspace uniqueless DB-less tests there, yes.
